qcodes/dataset/sqlite_base.py

```diff
@@ -1363,29 +1507,113 @@ def add_meta_data(conn: sqlite3.Connection,
+def get_run_timestamp_from_run_id(conn: SomeConnection,
+                                  run_id: int) -> float:
     return select_one_where(conn, "runs", "run_timestamp", "run_id", run_id)
+
+
+def update_GUIDs(conn: SomeConnection) -> None:
+    """
+    Update all GUIDs in this database where either the location code or the
+    work_station code is zero to use the location and work_station code from
+    the qcodesrc.json file in home. Runs where it is not true that both codes
+    are zero are skipped.
+    """
+
+    log.info('Commencing update of all GUIDs in database')
+
+    cfg = qc.Config()
+
+    try:
+        location = cfg['GUID_components']['location']
+    except KeyError:
+        log.warning('No location information found in the config. '
+                    'Can not proceed.')
+        return
+    try:
+        work_station = cfg['GUID_components']['work_station']
+    except KeyError:
+        log.warning('No work_station information found in the config. '
+                    'Can not proceed.')
+        return
+
+    if location == 0:
+        log.warning('The location is still set to the default (0). Can not '
+                    'proceed. Please configure the location before updating '
+                    'the GUIDs.')
+    if work_station == 0:
+        log.warning('The work_station is still set to the default (0). Can not'
+                    ' proceed. Please configure the location before updating '
+                    'the GUIDs.')
+
+    query = f"select MAX(run_id) from runs"
+    c = atomic_transaction(conn, query)
+    no_of_runs = c.fetchall()[0][0]
+
+    # now, there are four actions we can take
+
+    def _both_nonzero(run_id: int, *args) -> None:
+        log.info(f'Run number {run_id} already has a valid GUID, skipping.')
+
+    def _location_only_zero(run_id: int, *args) -> None:
+        log.warning(f'Run number {run_id} is has a zero (default) location '
+                    'code, but a non-zero work station code. Please manually '
+                    'resolve this, skipping the run now.')
+
+    def _workstation_only_zero(run_id: int, *args) -> None:
+        log.warning(f'Run number {run_id} is has a zero (default) work station'
+                    ' code, but a non-zero location code. Please manually '
+                    'resolve this, skipping the run now.')
+
+    def _both_zero(run_id: int, conn, guid_comps) -> None:
+        guid_str = generate_guid(timeint=guid_comps['time'],
+                                 sampleint=guid_comps['sample'])
+        with atomic(conn) as conn:
+            sql = f"""
+                   UPDATE runs
+                   SET guid = ?
+                   where run_id == {run_id}
+                   """
+            cur = conn.cursor()
+            cur.execute(sql, (guid_str,))
+
+        log.info(f'Succesfully updated run number {run_id}.')
+
+    actions: Dict[Tuple[bool, bool], Callable]
```

Comment:
`Callable[int, SomeConnection, Dict]`? or there is no reason to add this level of details?

Reply:
`Dict[Tuple[bool, bool], Callable]` is enough to make mypy shut up 😉